### ISSUE TITLE
KeyError: 'web' .. it doesn't make sense

### ISSUE THREAD

**Issue by jl2035** [odoo/odoo#25275]
I'm fighting with this problem for days now. I don't understand how is it possible that Odoo does not find the 'web' module. The files are there, nothing was moved or deleted.

None of these solutions worked for me:
- https://www.odoo.com/groups/community-59/community-24092362?mode=thread&date_begin=&date_end=
- https://stackoverflow.com/questions/39569445/how-to-fix-keyerror-web-in-odoo8
- https://www.odoo.com/forum/help-1/question/keyerror-web-odoo-10-127241

My traceback:

> Traceback (most recent call last):
  File "/home/odoo/venv/lib/python2.7/site-packages/werkzeug/serving.py", line 193, in run_wsgi
    execute(self.server.app)
  File "/home/odoo/venv/lib/python2.7/site-packages/werkzeug/serving.py", line 181, in execute
    application_iter = app(environ, start_response)
  File "/home/odoo/odoo10/odoo/service/server.py", line 249, in app
    return self.app(e, s)
  File "/home/odoo/odoo10/odoo/service/wsgi_server.py", line 184, in application
    return werkzeug.contrib.fixers.ProxyFix(application_unproxied)(environ, start_response)
  File "/home/odoo/venv/lib/python2.7/site-packages/werkzeug/contrib/fixers.py", line 152, in __call__
    return self.app(environ, start_response)
  File "/home/odoo/odoo10/odoo/service/wsgi_server.py", line 172, in application_unproxied
    result = handler(environ, start_response)
  File "/home/odoo/odoo10/odoo/http.py", line 1308, in __call__
    return self.dispatch(environ, start_response)
  File "/home/odoo/odoo10/odoo/http.py", line 1483, in dispatch
    ir_http.pool.signal_caches_change()
  File "/home/odoo/odoo10/addons/website/models/ir_http.py", line 218, in _dispatch
    resp = super(Http, cls)._dispatch()
  File "/home/odoo/odoo10/addons/web_editor/models/ir_http.py", line 21, in _dispatch
    return super(IrHttp, cls)._dispatch()
  File "/home/odoo/odoo10/odoo/addons/base/ir/ir_http.py", line 199, in _dispatch
    return cls._handle_exception(e)
  File "/home/odoo/odoo10/addons/website/models/ir_http.py", line 270, in _handle_exception
    return super(Http, cls)._handle_exception(exception)
  File "/home/odoo/odoo10/odoo/addons/base/ir/ir_http.py", line 169, in _handle_exception
    return request._handle_exception(exception)
  File "/home/odoo/odoo10/odoo/http.py", line 768, in _handle_exception
    return super(HttpRequest, self)._handle_exception(exception)
  File "/home/odoo/odoo10/odoo/addons/base/ir/ir_http.py", line 195, in _dispatch
    result = request.dispatch()
  File "/home/odoo/odoo10/odoo/http.py", line 827, in dispatch
    r = self._call_function(**self.params)
  File "/home/odoo/odoo10/odoo/http.py", line 333, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/home/odoo/odoo10/odoo/service/model.py", line 101, in wrapper
    return f(dbname, *args, **kwargs)
  File "/home/odoo/odoo10/odoo/http.py", line 326, in checked_call
    result = self.endpoint(*a, **kw)
  File "/home/odoo/odoo10/odoo/http.py", line 935, in __call__
    return self.method(*args, **kw)
  File "/home/odoo/odoo10/odoo/http.py", line 506, in response_wrap
    response = f(*args, **kw)
  File "/home/odoo/odoo10/addons/web/controllers/main.py", line 493, in load_locale
    addons_path = http.addons_manifest['web']['addons_path']
KeyError: 'web'`

Impacted version is: Odoo10

**Comment by sswapnesh** [comment]
Hello @jl2035  Could you please share your config file ?
Look like `web` is not loaded in to addons path you are given.

**Comment by gdgellatly** [comment]
Server wide modules. Google it. It's not that web isn't in the addons path.
It's that it isn't loaded.

**Comment by pedrobaeza** [comment]
Hi, this seems something related to your deployment, as all the recent builds are correctly deployed in runbot (http://runbot.odoo.com/runbot), so please be sure you are following the proper installation instructions. Check them on https://www.odoo.com/documentation/11.0/setup/install.html.

**Comment by jl2035** [comment]
@sswapnesh this is my config:

> [options]
; This is the password that allows database operations:
admin_passwd = *****
db_host = False
db_port = False
db_user = odoo
db_password = ********
dbfilter = ^%d$
addons_path = /home/odoo/odoo10/addons,/home/odoo/customaddons
proxy_mode = True
#workers = 3
workers = 0
limit_time_cpu = 60000
limit_time_real = 120000
#limit_memory_hard = 1605306368
logfile = /var/log/odoo/odoo.log

@gdgellatly what are you trying to say? can you be more specific? Why isn't the web module loaded? I've googled this problem for two days and I provided 3 links, that do not solve my problem..

**Comment by gdgellatly** [comment]
server_wide_modules = ['web', 'web_kanban']

**Comment by jl2035** [comment]
@gdgellatly sir, can you please add some context? how can this line of code help me?

**Comment by lukebranch** [comment]
@jl2035 ,

You need to add this line to your odoo config file (eg. /etc/odoo-server.conf). It is to load the module first, instance-wide for all databases in that specific Odoo instance.

Add this line to your Odoo config file first and restart to see if it resolves the problem. If not, the most likely cause is that you have some incorrectly written custom modules (developed by yourself or a third-party) which have improper inheritance due to incorrect dependencies in your module(s) __manifest__.py or __openerp__.py.

**Comment by jl2035** [comment]
@lukebranch Sadly adding the line does not solve the issue :(

I don't get it. How could there be a problem in a custom module. My instance was working perfectly fine, until I started getting this KeyError. And I didn't change any code or configuration. Even my nginx configuration is the same as before.

**Comment by lukebranch** [comment]
@jl2035 ,

I'd suggest posting a copy of your odoo-server.conf file here or provide a link to a gist containing your config (removing sensitive information like passwords, etc. first of course). I'd say you probably have a  community module somewhere causing this issue.

**Comment by jl2035** [comment]
@lukebranch I already did. Please scroll a few comments higher..

**Comment by gdgellatly** [comment]
@jl2035 I just hit this exact same error testing an automated build.  I got the syntax wrong, 

server_wide_modules = web,web_kanban

However in this case I was attempting to start 3 containers at once, all with a -d flag and all creating a database.  Needless to say postgres spat dummies everywhere.  So I had a half created database and no matter what I tried it wouldn't load, just had that error.  But in my case I could just drop the database and start again.

But after dropping the -d flag from 2 of the containers, and adding the above line to config the error was gone.

**Comment by jl2035** [comment]
I've corrected the syntax, but nothing has changed. Still this same stupid error.

This will probably be my main reason why I'll be moving away from Odoo. I can't believe how the instance can get so screwed up, only by itself (Nobody has changed any code). I cannot afford such failures, especially not on production servers.

Sad... I feel like I've just wasted the last 3 years of my life :(

**Comment by jl2035** [comment]
@moylop260 I'm running multiple applications/databases on multiple domains and subdomains, so my only possible dbfilter is:

`dbfilter = ^%d$`

I already tried deleting and recreating the database on one of these domains, but of course that wouldn't fix the problem. By the way, I'm getting this same error on all databases (even newly created with no additional modules installed), either when trying to login, or just by visiting the website (in case database has the Website app installed).

**Comment by moylop260** [comment]
My question is about test it one time...
I mean, if you confirm that the issue is fixed using one database you can have a clue about where looking for... (proxy, cache, sessions...)

**Comment by jl2035** [comment]
Using one database with dbfilter and db_name options does not solve the issue either. Where do I look now?

**Comment by gdgellatly** [comment]
Your issue is the database not the code. That should be obvious. If server
wide isn't fixing then its definitely an existing database of the odoo user.

If you run a

select res_id from ir_model_data where name = 'login' and module = 'web';
and get no result that's your problem

Also run select 1 from ir_ui_view where id = id from above; no result
that's your problem.

Do for each database. Find the culprit and just reinstall web from command
line or drop the db.

But I can assure someone has done something. If it's not you I'd be
checking that the database is not externally accessible as lots of postgres
hacks going around atm for badly deployed dbs that aren't up to date.

**Comment by jl2035** [comment]
I got the result with both commands on all databases.

But...

I've somehow fixed the problem. Apparently it was a git fuckup. I've cleared it up, pulled the 10.0 branch again and updated 'web' on all databases and the error is gone now, I think.

I've also realized why I didn't discover this before. I downloaded another copy of odoo to check if my code is screwed (yeah stupid). I've executed ./odoo-bin there, but I've provided the same config file, that points to my initial addons folder, which of course was not in the folder where my new copy of odoo was.

Thanks guys, without your hints I would simply give up.
